2021/BondielliCamarda/cusum.c

```diff
@@ -0,0 +1,213 @@
+
+    int ret = ndpi_cusum_init(&cusum, alpha, beta, sigma, 3);
+
+    printf("Start Alghoritm\n");
+
+    for (int i = 0; i < num; ++i) {
+        double prediction, threshold = 1500;
+        int rc = ndpi_cusum_add_value(&cusum, v[i], &prediction);
+
+        //it checks if the prediction computed goes beyond a given fixed threshold
+        if (prediction > threshold) printf("Attention: anomaly detected.\n");
+
+        printf("%2u)\t%12.3f\t%.3f\n",i, v[i], prediction);
+    }
+
+    return 0;
+}
+
+/*
+  Cumulative Sum (CUSUM) Algorithm
+*/
+
+//it initializes the parameters passed to the algorithm
+int ndpi_cusum_init(struct ndpi_cusum_struct *cusum, double alpha, double beta, double sigma, int window_size) {
+    
+    memset(cusum, 0, sizeof(struct ndpi_cusum_struct));
+
+    cusum->params.alpha = alpha;
+    cusum->params.beta = beta;
+
+    cusum->params.sigma = sigma;
+    cusum->square_sigma = sigma * sigma;
+
+    cusum->params.window_size = window_size;
+
+    cusum->num_values = 0;
+    cusum->last_forecast = 0;
+
+    cusum->values_window = (double*)malloc(cusum->params.window_size * sizeof(double));
+
+    cusum->start_cusum = false;
+
+    return(0);
+}
+
+/*
+   The CUSUM algorithm belongs to the family of change point detection algorithms that are based 
+   on hypothesis testing; it involves a cumulative sum of a statistical data records added to some
+   weights and when the sum exceeds a certain threshold an anomaly is detected.
+   
+   z_{n} = max( (z_{n-1} + (alpha*mu_{n-1} / sigma^{2}) * (x_{n} - mu_{n-1} - alpha*mu_{n-1}/2) , 0)
+   mu_{n} = beta*mu_{n-1} + (1- beta)*x_{n}
+   
+   Input:
+    cusum            Datastructure previously initialized
+    value            The value to add to the measurement
+   
+   Output:
+    forecast         The forecasted value
+   
+   Return code:
+    0                Too early: we don't have enough values. Output values are zero.
+    1                Normal processing: forecast is meaningful
+*/
+
+int ndpi_cusum_add_value(struct ndpi_cusum_struct *cusum, const u_int64_t _value, double *forecast) {
+    double value = (double)_value;
+    double z;
+    double mean_window;
+    double mu;
+    double alpha_mu;
+
+    //it checks if the algorithm has started or not
+    if (!cusum->start_cusum) {
+        if (cusum->num_values < cusum->params.window_size) { //algorithm not yet started
+            //adds value to the structure
+            cusum->values_window[cusum->num_values] = value;
+
+            z = 0;
+            cusum->num_values++;
+            *forecast = max(cusum->last_forecast + z, 0);
+
+            return 0;
+        } 
+        else { //cusum->num_values = cusum->params.window_size
+
+            z = 0;
+
+            //adds value to the structure
+            cusum->values_window[cusum->num_values % cusum->params.window_size] = value;
+
+            //last_mu = mean 
+            cusum->last_mu = mean(cusum->values_window, cusum->params.window_size);
```

Comment:
Questo calcolo viene fatto per ogni valore e va ottimizzato

Reply:
Abbiamo sostituito il calcolo della media, tenendo in una variabile la somma dei valori via via che vengono letti.